doc/sailfish-webview-qmozcontext.cpp

```diff
@@ -0,0 +1,246 @@
+/****************************************************************************************
+**
+** Copyright (c) 2021 Open Mobile Platform LLC
+** All rights reserved.
+**
+****************************************************************************************/
+
+#include "sailfish-webview-qmozcontext.h"
+
+// Used only for documentation purposes
+
+/*!
+    \class QMozContext
+    \brief Web engine context implementation
+    \inmodule SailfishWebView
+
+    \c QMozContext is inherited by \l WebEngine to provide methods for
+    interacting with the underlying EmbedLiteApp and gecko engine.
+
+    Developers should access the methods through \l WebEngine rather than
+    instantiating a \c QMozContext explicitly.
+*/
```

Comment:
Should these methods be described as part of `QMozContext` and inherited by `WebEngine`, or should we pretend they're just part of `WebEngine`? (I'm not actually sure if we can achieve the latter nicely with qdoc though).

Reply:
I'd document all the relevant API bits here.

Reply:
I just want to confirm in case I'm misunderstanding: document them here in a dummy `QMozContext` that `WebEngine` inherits?

Reply:
Yes, and methods/properites that I see relevant are:
```
properties:
initialized

methods (C++ only)
void addObservers(const std::vector<std::string> &aObserversList);
static QMozContext *instance();

slots
void addObserver(const QString &aTopic);
void removeObserver(const QString &aTopic);

void notifyObservers(const QString &topic, const QString &value);
void notifyObservers(const QString &topic, const QVariant &value);

void addComponentManifest(const QString &manifestPath);

void runEmbedding(int aDelay = -1);
void stopEmbedding();
void setProfile(const QString &);

signals:
void contextDestroyed();
void lastViewDestroyed();
void lastWindowDestroyed();
void recvObserve(const QString message, const QVariant data);
``` 

WebEngine does call runEmbedding from its initialization but custom initialization process would require setProfile, addComponentManifest, runEmbedding, and stopEmbedding. 

I.e. would leave out registerWindow, isAccelerated, GetApp, PostUITask, PostCompositorTask, CancelTask, setIsAccelerated, setViewCreator, createView, notifyFirstUIInitialized at least

Reply:
Thanks for the clarifications, that all makes sense. I've updated the PR so that the current roster looks like this. There are minor differences from the list you provided, but hopefully the differences are for valid reasons (e.g. `isInitialized` seems to be a consequence of having the `initialized` property). 
<h2 id="properties">Properties</h2>
<ul>
<li class="fn"><b>initialized</b> : const bool</li>
</ul>
<a name="public-functions"></a>
<h2 id="public-functions">Public Functions</h2>
<div class="table">
<br/>void <b>addObservers</b>(const std::vector&lt;std::string&gt; &amp;aObserversList)
<br/>bool <b>isInitialized</b>() const
<br/>void <b>removeObservers</b>(const std::vector&lt;std::string&gt; &amp;aObserversList)

</div>
<a name="public-slots"></a>
<h2 id="public-slots">Public Slots</h2>
<div class="table">
<br/>void <b>addComponentManifest</b>(const QString &amp;manifestPath)
<br/>void <b>addObserver</b>(const QString &amp;aTopic)
<br/>void <b>notifyObservers</b>(const QString &amp;topic, const QString &amp;value)
<br/>void <b>notifyObservers</b>(const QString &amp;topic, const QVariant &amp;value)
<br/>void <b>removeObserver</b>(const QString aTopic)
<br/>void <b>runEmbedding</b>(int aDelay = -1)
<br/>void <b>setProfile</b>(const QString &amp;profilePath)
<br/>void <b>stopEmbedding</b>()

</div>
<a name="signals"></a>
<h2 id="signals">Signals</h2>
<div class="table">
<br/>void <b>contextDestroyed</b>()
<br/>void <b>initialized</b>()
<br/>void <b>lastViewDestroyed</b>()
<br/>void <b>lastWindowDestroyed</b>()
<br/>void <b>recvObserve</b>(const QString message, const QVariant data)

</div>

Reply:
It seems strange to have an initialized() signal as well as an initialized property.  Should the signal be initializedChanged()?

Reply:
That's a good question; we had some discussion about it [here](https://git.sailfishos.org/mer-core/qtmozembed/merge_requests/95#note_89619). Use of `initialized()` for the signal predates the change to a property, so I kept it like that to avoid complications. It could be revisited. For QML Raine discovered that `onInitializedChanged` still works, but for the C++ side I guess the real signal name is needed.

Reply:
Originally I was thinking also ```initializedChanged``` but as generated slot (onInitializedChanged) is still there on QML due to signal name I think this is actually better. 

As this is singleton and should not be initialized explicitly following example is better than only having initializedChanged signal.

```
    Connections {
        target: WebEngine

        onInitialized: console.log("======== Do something")
```
At the same time, we have quite a few places where we are using ```onInitialized:``` and I do think that having WebEngine, WebView, WebEngineSettings in sync is the key factor and eases the API usage.

When going forward, at some point we could consider adding ```WebComponent``` component and attached signals for  initialized and destroyed. Something similar to ```Component.onCompleted``` => ```WebComponent.onInitialized``` => Out of scope for now I think.